sw/host/opentitanlib/src/debug/dmi.rs

```diff
@@ -0,0 +1,353 @@
+        ensure!(idx < 12, "data register index out of range {:#x}", idx);
+        self.dmi_write(DATA0 + idx, data)
+    }
+}
+
+/// A DMI debugger with specific hart selected.
+pub struct DmiHart<'a, D> {
+    debugger: &'a mut DmiDebugger<D>,
+
+    /// The value of DMCONTROL with hasel, hartsello and hartselhi set.
+    hart_select: u32,
+}
+
+impl<D> Deref for DmiHart<'_, D> {
+    type Target = DmiDebugger<D>;
+
+    fn deref(&self) -> &Self::Target {
+        self.debugger
+    }
+}
+
+impl<D> DerefMut for DmiHart<'_, D> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.debugger
+    }
+}
+
+/// State of the hart.
+///
+/// If both `running` and `halted` are false, then the hart is in the process of transitioning between
+/// the two states (i.e. resuming or halting).
+pub struct HartState {
+    pub running: bool,
+    pub halted: bool,
+}
+
+impl<D: Dmi> DmiHart<'_, D> {
+    /// Read `dmstatus` for the selected hart.
+    pub fn dmstatus(&mut self) -> Result<u32> {
+        let dmstatus = self.debugger.dmi_read(DMSTATUS)?;
+
+        // `dmstatus` register have fields for any hart and all harts. If only a single hart is selected,
+        // then the "all" and "any" values should match. This performs a sanity check.
+        if (dmstatus ^ (dmstatus >> 1))
+            & (DMSTATUS_ANYHALTED_MASK
+                | DMSTATUS_ANYRUNNING_MASK
+                | DMSTATUS_ANYUNAVAIL_MASK
+                | DMSTATUS_ANYNONEXISTENT_MASK
+                | DMSTATUS_ANYRESUMEACK_MASK
+                | DMSTATUS_ANYHAVERESET_MASK)
+            != 0
+        {
+            bail!(
+                "Invalid dmstatus {:#x}: any and all bits mismatch",
+                dmstatus
+            );
+        }
+
+        Ok(dmstatus)
+    }
+
+    /// Write to dmcontrol without affecting hart selection.
+    pub fn set_dmcontrol(&mut self, value: u32) -> Result<()> {
+        self.debugger.dmi_write(DMCONTROL, value | self.hart_select)
+    }
+
+    /// Read hart info of the selected hart.
+    pub fn hartinfo(&mut self) -> Result<u32> {
+        self.debugger.dmi_read(HARTINFO)
+    }
+
+    /// Return the state of the hart.
+    pub fn state(&mut self) -> Result<HartState> {
+        let dmstatus = self.dmstatus()?;
+        let running = dmstatus & DMSTATUS_ANYRUNNING_MASK != 0;
+        let halted = dmstatus & DMSTATUS_ANYHALTED_MASK != 0;
+        assert!(!(running && halted));
+        Ok(HartState { running, halted })
+    }
+
+    /// Set the halt request bit.
+    pub fn set_halt_request(&mut self, active: bool) -> Result<()> {
+        self.set_dmcontrol(if active { DMCONTROL_HALTREQ_MASK } else { 0 })
+    }
+
+    /// Wait for the hart to halt.
+    pub fn wait_halt(&mut self) -> Result<()> {
+        // Per RISC-V debug specification, harts must respond within 1 second of receiving a halt or
+        // resume request.
+        poll_until(Duration::from_secs(1), Duration::from_secs(1), || {
```

Comment:
Should we reduce the delay here to `Duration::from_millis(100)` or something? Currently, won't this poll at 0 and 1 seconds but not inbetween?

Reply:
Nice catch!